File: src/server/net/net.cc
/*
 * Copyright (C) 2017, Georg Kotheimer <georg.kotheimer@mailbox.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <base/Common.h>
#include <base/log/Services.h>
#include <base/util/Profile.h>
#include <base/Panic.h>

#include <m3/net/Net.h>
#include <m3/server/RequestHandler.h>
#include <m3/server/Server.h>
#include <m3/stream/Standard.h>
#include <m3/vfs/VFS.h>
#include <pci/Device.h>
#include <thread/ThreadManager.h>

#include "driver/driver.h"

#include "lwipopts.h"
#include "lwip/sys.h"
#include "lwip/init.h"
#include "lwip/netif.h"
#include "lwip/ip_addr.h"
#include "lwip/pbuf.h"
#include "lwip/prot/ethernet.h"
#include "lwip/etharp.h"
#include "lwip/timeouts.h"

#include "sess/SocketSession.h"

#include <assert.h>
#include <m3/session/NetworkManager.h>
#include <m3/session/ServerSession.h>
#include <queue>
#include <cstring>
#include <memory>

using namespace net;
using namespace m3;

class NMRequestHandler;
using net_reqh_base_t = RequestHandler<
    NMRequestHandler, NetworkManager::Operation, NetworkManager::COUNT, NMSession
>;

static Server<NMRequestHandler> *srv;

class NMRequestHandler: public net_reqh_base_t {
public:
    explicit NMRequestHandler(WorkLoop *wl, NetDriver *driver)
        : net_reqh_base_t(),
          _wl(wl),
          _driver(driver),
          _rgate(RecvGate::create(nextlog2<32 * NMSession::MSG_SIZE>::val, nextlog2<NMSession::MSG_SIZE>::val)) {
        add_operation(NetworkManager::CREATE, &NMRequestHandler::create);
        add_operation(NetworkManager::BIND, &NMRequestHandler::bind);
        add_operation(NetworkManager::LISTEN, &NMRequestHandler::listen);
        add_operation(NetworkManager::CONNECT, &NMRequestHandler::connect);
        add_operation(NetworkManager::CLOSE, &NMRequestHandler::close);
        add_operation(NetworkManager::STAT, &NMRequestHandler::stat);
        add_operation(NetworkManager::SEEK, &NMRequestHandler::seek);
        add_operation(NetworkManager::NEXT_IN, &NMRequestHandler::next_in);
        add_operation(NetworkManager::NEXT_OUT, &NMRequestHandler::next_out);
        add_operation(NetworkManager::COMMIT, &NMRequestHandler::commit);

        using std::placeholders::_1;
        _rgate.start(wl, std::bind(&NMRequestHandler::handle_message, this, _1));
    }

    virtual Errors::Code open(NMSession **sess, capsel_t srv_sel, const String &) override {
        *sess = new SocketSession(_wl, srv_sel, _rgate);
        _sessions.append(*sess);
        return Errors::NONE;
    }

    virtual Errors::Code obtain(NMSession *sess, KIF::Service::ExchangeData &data) override {
        return sess->obtain(srv->sel(), data);
    }

    virtual Errors::Code delegate(NMSession *sess, KIF::Service::ExchangeData &data) override {
        return sess->delegate(data);
    }

    virtual Errors::Code close(NMSession *sess) override {
        if(sess->type() == NMSession::SOCKET)
            _sessions.remove(sess);
        delete sess;
        _rgate.drop_msgs_with(reinterpret_cast<label_t>(sess));
        return Errors::NONE;
    }

    virtual void shutdown() override {
        // delete sessions to remove items from workloop etc.
        for(auto it = _sessions.begin(); it != _sessions.end(); ) {
            auto old = it++;
            delete &*old;
        }

        _driver->stop();
        _rgate.stop();
    }

    void create(GateIStream &is) {
        NMSession *sess = is.label<NMSession*>();
        sess->create(is);
    }

    void bind(GateIStream &is) {
        NMSession *sess = is.label<NMSession*>();
        sess->bind(is);
    }

    void listen(GateIStream &is) {
        NMSession *sess = is.label<NMSession*>();
        sess->listen(is);
    }

    void connect(GateIStream &is) {
        NMSession *sess = is.label<NMSession*>();
        sess->connect(is);
    }

    void close(GateIStream &is) {
        NMSession *sess = is.label<NMSession*>();
        sess->close(is);
        // we don't get the close call from our resource manager for files (child sessions)
        if(sess->type() == NMSession::FILE)
            close(sess);
    }

    void next_in(GateIStream &is) {
        NMSession *sess = is.label<NMSession*>();
        sess->next_in(is);
    }

    void next_out(GateIStream &is) {
        NMSession *sess = is.label<NMSession*>();
        sess->next_out(is);
    }

    void commit(GateIStream &is) {
        NMSession *sess = is.label<NMSession*>();
        sess->commit(is);
    }

    void seek(GateIStream &is) {
        NMSession *sess = is.label<NMSession*>();
        sess->seek(is);
    }

    void stat(GateIStream &is) {
        NMSession *sess = is.label<NMSession*>();
        sess->stat(is);
    }

private:
    WorkLoop *_wl;
    NetDriver *_driver;
    RecvGate _rgate;
    SList<NMSession> _sessions;
};

static std::queue<struct pbuf*> recvQueue;

static bool eth_alloc_callback(void *&pkt, void *&buf, size_t &buf_size, size_t size) {
    /* Allocate pbuf from pool */
    struct pbuf *p = pbuf_alloc(PBUF_RAW, size, PBUF_POOL);

    if(p != NULL) {
        pkt = p;
        buf = p->payload;
        buf_size = p->len;
        return true;
    } else
        return false;
}

static void eth_next_buf_callback(void *&pkt, void *&buf, size_t &buf_size) {
    struct pbuf *p = static_cast<struct pbuf *>(pkt)->next;
    pkt = p;
    buf = p ? p->payload : 0;
    buf_size = p ? p->len : 0;
}

static void eth_recv_callback(void *pkt) {
    struct pbuf *p = static_cast<pbuf *>(pkt);

    /* Put in a queue which is processed in main loop */
    // TODO: Get rid of the queue or at least limit it in size.
    recvQueue.push(p);
}

static err_t netif_output(struct netif *netif, struct pbuf *p) {
    LINK_STATS_INC(link.xmit);

    SLOG(NET, "netif_output with size " << p->len);

    if(p->next) {
        // TODO: Use copy callback instead of scratch buffer
        SLOG(NET, "netif_output: Using scratch buffer for pbuf chain.");
        uint8_t *pkt = (uint8_t*)malloc(p->tot_len);
        if(!pkt) {
            SLOG(NET, "Not enough memory to read packet");
            return ERR_MEM;
        }

        pbuf_copy_partial(p, pkt, p->tot_len, 0);
        NetDriver *driver = static_cast<NetDriver*>(netif->state);
        if(!driver->send(pkt, p->tot_len)) {
            free(pkt);
            SLOG(NET, "netif_output failed!");
            return ERR_IF;
        }

        free(pkt);
    } else {
        NetDriver *driver = static_cast<NetDriver*>(netif->state);
        if(!driver->send(p->payload, p->tot_len)) {
            SLOG(NET, "netif_output failed!");
            return ERR_IF;
        }
    }

    return ERR_OK;
}

static void netif_status_callback(struct netif *netif) {
    SLOG(NET, "netif status changed " << ipaddr_ntoa(netif_ip4_addr(netif))
        << " to " << ((netif->flags & NETIF_FLAG_UP) ? "up" : "down"));
}

static err_t netif_init(struct netif *netif) {
    NetDriver *driver = static_cast<NetDriver*>(netif->state);

    netif->linkoutput = netif_output;
    netif->output = etharp_output;
    // netif->output_ip6 = ethip6_output;
    netif->mtu = 1500;
    netif->flags = NETIF_FLAG_BROADCAST | NETIF_FLAG_ETHARP | NETIF_FLAG_ETHERNET |
                   NETIF_FLAG_IGMP | NETIF_FLAG_MLD6;

    m3::net::MAC mac = driver->readMAC();
    static_assert(m3::net::MAC::LEN == sizeof(netif->hwaddr), "mac address size mismatch");
    SMEMCPY(netif->hwaddr, mac.bytes(), m3::net::MAC::LEN);
    netif->hwaddr_len = sizeof(netif->hwaddr);

    return ERR_OK;
}

static bool link_state_changed(struct netif *netif) {
    return static_cast<NetDriver *>(netif->state)->linkStateChanged();
}

static bool link_is_up(struct netif *netif) {
    return static_cast<NetDriver *>(netif->state)->linkIsUp();
}

int main(int argc, char **argv) {
    if(argc != 4)
        exitmsg("Usage: " << argv[0] << " <name> <ip address> <netmask>");

    ip_addr_t ip;
    if(!ipaddr_aton(argv[2], &ip))
        exitmsg(argv[2] << " is not a well formed ip address.");
    else
        SLOG(NET, "ip: " << ipaddr_ntoa(&ip));

    ip_addr_t netmask;
    if(!ipaddr_aton(argv[3], &netmask))
        exitmsg(argv[3] << " is not a well formed netmask.");
    else SLOG(NET, "netmask: " << ipaddr_ntoa(&netmask));

    struct netif netif;

    WorkLoop wl;

    NetDriver *driver = NetDriver::create(argv[1], &wl, eth_alloc_callback, eth_next_buf_callback,
                                          eth_recv_callback);

    lwip_init();

    netif_add(&netif, &ip, &netmask, IP4_ADDR_ANY, driver, netif_init, netif_input);
    netif.name[0] = 'e';
    netif.name[1] = '0';
    // netif_create_ip6_linklocal_address(&netif, 1);
    // netif.ip6_autoconfig_enabled = 1;
    netif_set_status_callback(&netif, netif_status_callback);
    netif_set_default(&netif);
    netif_set_up(&netif);

    /* Start DHCP */
    // dhcp_start(&netif );
    srv = new Server<NMRequestHandler>(argv[1], &wl, new NMRequestHandler(&wl, driver));

    while(wl.has_items()) {
        /* Check link state, e.g. via MDIO communication with PHY */
        if(link_state_changed(&netif)) {
            if(link_is_up(&netif))
                netif_set_link_up(&netif);
            else
                netif_set_link_down(&netif);
        }

        /* Check for received frames, feed them to lwIP */
        size_t maxReceiveCount = SocketSession::MAX_SEND_RECEIVE_BATCH_SIZE;
        while(!recvQueue.empty() && maxReceiveCount--) {
            struct pbuf *p = recvQueue.front();
            recvQueue.pop();
            LINK_STATS_INC(link.recv);

            err_t err = netif.input(p, &netif);
            if(err != ERR_OK) {
                SLOG(NET, "netif.input() failed with error " << err << ", dropping packet!");
                pbuf_free(p);
            }
        }

        /* Cyclic lwIP timers check */
        sys_check_timeouts();

        // Hack: run the workloop manually
        // - interrupt receive gate
        wl.tick();

        // TODO: Consider FileSession::handle_..., More?
        // Sleep according to sys_timeouts_sleeptime() if there is nothing to do.
        if(recvQueue.empty()) {
            if(auto sleep_ms = sys_timeouts_sleeptime()) {
                cycles_t sleep_time = sleep_ms * (m3::DTU::get().clock() / 1000);
                cycles_t start = m3::DTU::get().tsc();
                SLOG(NET_ALL, "@" << start << " try_sleep: " << sleep_time << " cycles"
                    << " (" << sleep_ms << " ms)");

                DTU::get().try_sleep(false, sleep_time);

                cycles_t stop = m3::DTU::get().tsc();
                SLOG(NET_ALL, "@" << stop << " wakeup: " << stop - start << " cycles"
                    << " (" << ((stop - start) * 1000 / m3::DTU::get().clock()) << " ms)");
            }
        }
    }

    delete srv;
    delete driver;

    return 0;
}

File: src/server/net/src/net.rs
/*
 * Copyright (C) 2021-2022 Nils Asmussen, Barkhausen Institut
 * Copyright (C) 2021, Tendsin Mende <tendsin.mende@mailbox.tu-dresden.de>
 * Copyright (C) 2017, Georg Kotheimer <georg.kotheimer@mailbox.tu-dresden.de>
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#![no_std]

use core::str::FromStr;

use m3::cap::Selector;
use m3::cell::{LazyStaticCell, StaticRefCell};
use m3::col::{BTreeMap, String, ToString, Vec};
use m3::com::{opcodes, GateIStream};
use m3::errors::{Code, Error};
use m3::io::LogFlags;
use m3::net::{log_net, NetLogEvent};
use m3::server::{
    CapExchange, ExcType, Handler, RequestHandler, Server, SessId, SessionContainer,
    DEF_MAX_CLIENTS,
};
use m3::tiles::OwnActivity;
use m3::time::{TimeDuration, TimeInstant};
use m3::{env, reply_vmsg};
use m3::{log, println};

use smoltcp::iface::{InterfaceBuilder, NeighborCache, Routes, SocketHandle};
use smoltcp::wire::{EthernetAddress, IpAddress, Ipv4Cidr};

use crate::driver::DriverInterface;
use crate::sess::SocketSession;
use crate::smoltcpif::socket::to_m3_addr;

mod driver;
mod ports;
mod sess;
mod smoltcpif;

const MAX_SOCKETS: usize = 64;
const MSG_SIZE: usize = 128;

static OWN_IP: LazyStaticCell<IpAddress> = LazyStaticCell::default();
static NAMESERVER: LazyStaticCell<IpAddress> = LazyStaticCell::default();
static OWN_MAC: [u8; 6] = [0x00, 0x0A, 0x35, 0x03, 0x02, 0x03];
static TIMEOUTS: StaticRefCell<Vec<(SocketHandle, TimeInstant)>> = StaticRefCell::new(Vec::new());

pub fn add_timeout(handle: SocketHandle, timeout: TimeInstant) {
    TIMEOUTS.borrow_mut().push((handle, timeout));
}

pub fn remove_timeout(handle: SocketHandle) {
    TIMEOUTS.borrow_mut().retain(|t| t.0 != handle);
}

fn next_timeout() -> Option<TimeInstant> {
    TIMEOUTS
        .borrow()
        .iter()
        .min_by(|a, b| a.1.cmp(&b.1))
        .map(|t| t.1)
}

struct NetHandler<'a> {
    reqhdl: RequestHandler<SocketSession, opcodes::Net>,
    // holds all the actual smoltcp sockets. Used for polling events on them.
    iface: DriverInterface<'a>,
}

impl Handler<SocketSession> for NetHandler<'_> {
    fn sessions(&mut self) -> &mut SessionContainer<SocketSession> {
        self.reqhdl.sessions()
    }

    fn open(
        &mut self,
        crt: usize,
        srv_sel: Selector,
        arg: &str,
    ) -> Result<(Selector, SessId), Error> {
        self.reqhdl.open(crt, srv_sel, arg)
    }

    fn exchange(
        &mut self,
        crt: usize,
        sid: SessId,
        xchg: &mut CapExchange<'_>,
        obtain: bool,
    ) -> Result<(), Error> {
        let Self { reqhdl, iface } = self;

        reqhdl.handle_capxchg_with(crt, sid, xchg, obtain, |reqhdl, opcode, ty, xchg| {
            assert!(opcode == opcodes::Net::Create.into());

            let sess = reqhdl
                .clients_mut()
                .get_mut(sid)
                .ok_or_else(|| Error::new(Code::InvArgs))?;
            match ty {
                ExcType::Obt(_) => sess.create_socket(xchg, iface),
                ExcType::Del(_) => Err(Error::new(Code::InvArgs)),
            }
        })
    }

    fn close(&mut self, crt: usize, sid: SessId) {
        if let Some(s) = self.reqhdl.clients_mut().get_mut(sid) {
            s.abort_all(&mut self.iface).unwrap();
        }

        self.reqhdl.clients_mut().remove(crt, sid);
    }
}

impl NetHandler<'_> {
    fn fetch_and_handle(&mut self) {
        let Self { reqhdl, iface } = self;

        reqhdl.fetch_and_handle_msg_with(|_, opcode, sess, is| match opcode {
            o if o == opcodes::Net::Bind.into() => sess.bind(is, iface),
            o if o == opcodes::Net::Listen.into() => sess.listen(is, iface),
            o if o == opcodes::Net::Connect.into() => sess.connect(is, iface),
            o if o == opcodes::Net::Abort.into() => sess.abort(is, iface),
            o if o == opcodes::Net::GetIP.into() => Self::get_ip(is),
            o if o == opcodes::Net::GetNameSrv.into() => Self::get_nameserver(is),
            _ => Err(Error::new(Code::InvArgs)),
        });
    }

    fn get_ip(is: &mut GateIStream<'_>) -> Result<(), Error> {
        let addr = to_m3_addr(OWN_IP.get());
        reply_vmsg!(is, Code::Success, addr.0)
    }

    fn get_nameserver(is: &mut GateIStream<'_>) -> Result<(), Error> {
        if !NAMESERVER.is_some() {
            return Err(Error::new(Code::NotSup));
        }

        let addr = to_m3_addr(NAMESERVER.get());
        reply_vmsg!(is, Code::Success, addr.0)
    }

    // processes outgoing events to clients
    fn process_outgoing(&mut self) -> bool {
        let iface = &mut self.iface;
        let mut res = false;
        self.reqhdl.clients_mut().for_each(|s| {
            res |= s.process_outgoing(iface);
        });
        res
    }

    // processes incoming events from clients and returns whether there is still work to do
    fn process_incoming(&mut self) -> bool {
        let iface = &mut self.iface;
        let mut res = false;
        self.reqhdl.clients_mut().for_each(|s| {
            res |= s.process_incoming(iface);
        });
        res
    }
}

pub fn own_ip() -> IpAddress {
    OWN_IP.get()
}

#[derive(Clone, Debug)]
pub struct NetSettings {
    driver: String,
    name: String,
    ip: smoltcp::wire::Ipv4Address,
    netmask: smoltcp::wire::Ipv4Address,
    nameserver: Option<smoltcp::wire::Ipv4Address>,
    gateway: Option<smoltcp::wire::Ipv4Address>,
    max_clients: usize,
}

impl Default for NetSettings {
    fn default() -> Self {
        NetSettings {
            driver: String::from("default"),
            name: String::default(),
            netmask: smoltcp::wire::Ipv4Address::new(255, 255, 255, 0),
            ip: smoltcp::wire::Ipv4Address::default(),
            nameserver: None,
            gateway: None,
            max_clients: DEF_MAX_CLIENTS,
        }
    }
}

fn usage() -> ! {
    println!(
        "Usage: {} [-d <driver>] [-m <max-clients>] [-a <netmask>] [-n <nameserver>] [-g <gateway>] <name> <ip>",
        env::args().next().unwrap()
    );
    println!();
    println!("  -d: the driver to use (lo=loopback or default=E1000/Fifo)");
    println!("  -m: the maximum number of clients (receive slots)");
    println!("  -a: the network mask to use (default: 255.255.255.0)");
    println!("  -n: the IP address of the DNS server");
    println!("  -g: the IP address of the default gateway");
    OwnActivity::exit_with(Code::InvArgs);
}

fn parse_args() -> Result<NetSettings, String> {
    let mut settings = NetSettings::default();

    let args: Vec<&str> = env::args().collect();
    let mut i = 1;
    while i < args.len() {
        match args[i] {
            "-m" => {
                settings.max_clients = args[i + 1]
                    .parse::<usize>()
                    .map_err(|_| String::from("Failed to parse client count"))?;
                i += 1;
            },
            "-d" => {
                settings.driver = args[i + 1].to_string();
                i += 1;
            },
            "-a" => {
                settings.netmask = smoltcp::wire::Ipv4Address::from_str(
                    args.get(i + 1).expect("Failed to read netmask!"),
                )
                .expect("Failed to parse netmask!");
                i += 1;
            },
            "-n" => {
                settings.nameserver = Some(
                    smoltcp::wire::Ipv4Address::from_str(
                        args.get(i + 1).expect("Failed to read nameserver!"),
                    )
                    .expect("Failed to parse nameserver IP!"),
                );
                i += 1;
            },
            "-g" => {
                settings.gateway = Some(
                    smoltcp::wire::Ipv4Address::from_str(
                        args.get(i + 1).expect("Failed to read gateway!"),
                    )
                    .expect("Failed to parse gateway IP!"),
                );
                i += 1;
            },
            _ => break,
        }
        i += 1;
    }

    if args.len() < i + 2 {
        usage();
    }

    settings.name = args.get(i).expect("Failed to read name!").to_string();
    settings.ip =
        smoltcp::wire::Ipv4Address::from_str(args.get(i + 1).expect("Failed to read ip!"))
            .expect("Failed to parse IP address!");
    Ok(settings)
}

#[no_mangle]
pub fn main() -> Result<(), Error> {
    smoltcpif::logger::init().unwrap();

    let settings = parse_args().unwrap_or_else(|e| {
        println!("Invalid arguments: {}", e);
        usage();
    });

    let mut neighbor_cache_entries = [None; 8];
    let neighbor_cache = NeighborCache::new(&mut neighbor_cache_entries[..]);

    let ip_cidr = smoltcp::wire::IpCidr::Ipv4(
        Ipv4Cidr::from_netmask(settings.ip, settings.netmask)
            .expect("Invalid IP-address/netmask pair"),
    );
    let ip_addr = ip_cidr.address();
    OWN_IP.set(ip_addr);

    if let Some(ns) = settings.nameserver {
        let ns_cidr =
            Ipv4Cidr::from_netmask(ns, settings.netmask).expect("Invalid nameserver/netmask pair");
        NAMESERVER.set(IpAddress::Ipv4(ns_cidr.address()));
    }

    let mut routes = Routes::new(BTreeMap::new());
    if let Some(gw) = settings.gateway {
        routes
            .add_default_ipv4_route(gw)
            .expect("Cannot add default route");
    }

    ports::init(MAX_SOCKETS);

    let iface = if settings.driver == "lo" {
        driver::DriverInterface::Lo(
            InterfaceBuilder::new(
                smoltcp::phy::Loopback::new(smoltcp::phy::Medium::Ethernet),
                Vec::with_capacity(MAX_SOCKETS),
            )
            .hardware_addr(EthernetAddress::from_bytes(&OWN_MAC).into())
            .neighbor_cache(neighbor_cache)
            .ip_addrs([ip_cidr])
            .routes(routes)
            .finalize(),
        )
    }
    else {
        #[cfg(target_vendor = "gem5")]
        let device = driver::E1000Device::new().expect("Failed to create E1000 driver");
        #[cfg(any(target_vendor = "hw", target_vendor = "hw22"))]
        let device = driver::AXIEthDevice::new().expect("Failed to create AXI ethernet driver");
        driver::DriverInterface::Eth(
            InterfaceBuilder::new(device, Vec::with_capacity(MAX_SOCKETS))
                .hardware_addr(EthernetAddress::from_bytes(&OWN_MAC).into())
                .neighbor_cache(neighbor_cache)
                .ip_addrs([ip_cidr])
                .routes(routes)
                .finalize(),
        )
    };

    let mut handler = NetHandler {
        reqhdl: RequestHandler::new_with(settings.max_clients, MSG_SIZE, 1)
            .expect("Unable to create request handler"),
        iface,
    };

    let serv = Server::new(&settings.name, &mut handler).expect("Failed to create server!");

    log!(
        LogFlags::Info,
        concat!(
            "netrs: created service {} with {{\n",
            "  driver={},\n",
            "  ip={:?},\n",
            "  nameserver={:?},\n",
            "  gateway={:?},\n",
            "}}"
        ),
        settings.name,
        settings.driver,
        settings.ip,
        settings.nameserver,
        settings.gateway,
    );

    let start = TimeInstant::now();

    'outer: loop {
        let sleep_nanos = loop {
            if serv.fetch_and_handle(&mut handler).is_err() {
                break 'outer;
            }

            // Check if we got some messages through our main rgate.
            handler.fetch_and_handle();

            // receive events from clients and push data to send into smoltcp sockets
            let sends_pending = handler.process_incoming();

            let cur_time = smoltcp::time::Instant::from_millis(start.elapsed().as_millis() as i64);

            // now poll smoltcp to send and receive packets
            if let Err(e) = handler.iface.poll(cur_time) {
                log!(LogFlags::NetPoll, "netrs: poll failed: {}", e);
            }

            // check for outgoing events we have to send to clients
            let recvs_pending = handler.process_outgoing();

            if !sends_pending && !recvs_pending && !handler.iface.needs_poll() {
                // ask smoltcp how long we can sleep
                match handler.iface.poll_delay(cur_time) {
                    // we need to call it again immediately => continue the loop
                    Some(d) if d.total_millis() == 0 => continue,
                    // we should not wait longer than `n` => sleep for `n`
                    Some(n) => break TimeDuration::from_millis(n.total_millis()),
                    // smoltcp has nothing to do => sleep until the next TCU message arrives
                    None => break TimeDuration::MAX,
                }
            }
        };

        let now = TimeInstant::now();
        let sleep_nanos = match next_timeout() {
            Some(timeout) if timeout > now && timeout - now < sleep_nanos => timeout - now,
            _ => sleep_nanos,
        };

        log_net(NetLogEvent::StartedWaiting, 0, 0);
        log!(LogFlags::NetPoll, "Sleeping for {:?}", sleep_nanos);
        OwnActivity::sleep_for(sleep_nanos).ok();
        log_net(NetLogEvent::StoppedWaiting, 0, 0);
    }

    Ok(())
}
